Add agroal dataSourceImplementation mode selection

zhouzhou19950825 opened the pull request:
Hello,
AgroalDataSourceConfigurationSupplier uses DataSourceImplementation.AGROAL to initialize dataSourceImplementation by default. I hope to have more diversified choices through configuration.
Through JMH's test performance comparison, Hikari's performance is far better than agroal, and agroal also supports the packaging of Hikari-related packages, so I want to add dataSourceImplementation parameters in DataSourceJdbcRuntimeConfig,
"agroal", "agroal_poolless" and "hikari" three parameters to select different data pools. PoolingEnable in DataSourceJdbcRuntimeConfig can also remove this parameter if it is not used elsewhere.

jmh test result: [https://github.com/zhouzhou19950825/somedemo/blob/main/jmhtest.jpg](url)

hope to be adopted，Thanks！

zhouzhou19950825 committed:
add agroal dataSourceImplementation

gastaldi commented on extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/DataSourceJdbcRuntimeConfig.java:
```diff
@@ -145,4 +145,10 @@ public class DataSourceJdbcRuntimeConfig {
+     * choose agroal DataSourceImplementation
+     */
+    @ConfigItem
+    public Optional<String> dataSourceImplementation = Optional.empty();
```
Make it `Optional<DataSourceImplementation>` instead? Then you can get rid of the if-else statements

zhouzhou19950825 replied:
That's great, it scales even better

zhouzhou19950825 replied:
Then I modify the code and upload again.
Thanks!

gastaldi commented on extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/DataSources.java:
```diff
@@ -209,11 +209,25 @@ public AgroalDataSource doCreateDataSource(String dataSourceName) {
         }
 
+        // Set pool-less mode poolingEnabled no need
+//        if (!dataSourceJdbcRuntimeConfig.poolingEnabled) {
```
This property still exists and should not be ignored

zhouzhou19950825 committed:
add field  dataSourceImplementation

add field Optional<DataSourceImplementation> dataSourceImplementation

zhouzhou19950825 committed:
choose agroal DataSourceImplementation

if(dataSourceJdbcRuntimeConfig.dataSourceImplementation.isPresent()){
            dataSourceConfiguration.dataSourceImplementation(dataSourceJdbcRuntimeConfig.dataSourceImplementation.get());
        }

Sanne commented:
@zhouzhou19950825 thanks for this, we can certainly experiment with the idea but I share @gsmet 's concerns.

I do love more performance but we shouldn't have multiple options to do the same thing, that's contrary to the spirit of an opinionated framework - and not ideal for us being able to maintain it all properly.

I'm skeptical about the claims of HicariCP being "much faster" in real world scenarios and all cases we need - let's wait for @barreiro 's opinion. Did you run the comparisons @barreiro suggestion during our previous chat? Would you mind sharing your findings?

zhouzhou19950825 commented:
@gsmet  agroal currently supports hikari, so this configuration will not affect. 
Moreover, the agroal specification can be expanded to access more connection pools, such as c3p0, druid, etc., which can be seamlessly slanted with quarkus-argoal and has strong scalability.
https://github.com/agroal/agroal

barreiro commented:
hi @zhouzhou19950825 Thanks for the PR.

I see 2 problems with it, both already mentioned. 

The first is that the Hikari implementation only support a subset of all the features in the Agroal API. Things like tx integration, xa, metrics, interception, etc. It's just a bare bones implementation for development purposes. The fact that Hikari does not provide these features was one of the reasons we started Agroal in the first place.

Second, the performance of Agroal should be comparable with Hikari. Also, both use similar strategies to deal with contention, so in a real world application the difference should be small. I will look to recreate the results you got on the benchmark, but keep in mind that you provided a single data point. Any performance analysis should include several scenarios in terms of resources, pool size and contention.

zhouzhou19950825 commented:
hi @barreiro,
 
Argoal already supports these modes, including Hikari: https://github.com/agroal/agroal/blob/master/agroal-api/src/main/java/io/agroal/api/configuration/AgroalDataSourceConfiguration.java
So I think it can seamlessly cross the street.

Looking forward to your JMH test report very much.

Thaks!

gsmet commented:
Given @barreiro is the developer of Agroal, I'm pretty sure he's aware of the limitations of Agroal with HikariCP.

I'm going to close this for now. Thanks for the effort anyway.
